### ISSUE TITLE
Add optional PEFT LoRA-FA optimizer support

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#725]
## Motivation

PEFT exposes LoRA-FA, an optimizer that freezes LoRA A matrices and trains B
matrices. Freezing A avoids retaining the input activation needed for A's gradient,
so the adapter-rank contribution to activation memory can be substantially lower.

Primary references:

- PEFT optimizer API: `create_lorafa_optimizer`
- LoRA-FA paper: https://arxiv.org/abs/2308.03303

## Local validation

With PEFT 0.20.0, a tiny real `PeftModel` confirmed that
`create_lorafa_optimizer()` creates a `LoraFAOptimizer`, sets every `lora_A`
parameter to `requires_grad=False`, and leaves `lora_B` trainable.

An isolated activation-saved-tensor probe for `X @ A @ B` used
`X: 512x4096` and ranks 16/64/256:

| Rank | Ordinary LoRA retained | Frozen-A retained |
|---:|---:|---:|
| 16 | 8.03125 MiB | 32 KiB |
| 64 | 8.125 MiB | 128 KiB |
| 256 | 8.5 MiB | 512 KiB |

The difference was exactly 8 MiB in each case: the storage of `X` in this toy
calculation. Input and B gradients matched for the same calculation, while A had no
gradient by design.

These values are **not** total or peak LLM VRAM savings, an end-to-end throughput
result, or a quality claim. Other model operations may still retain the same input,
and backward recomputation/optimizer state are outside this micro-test.

## Proposed scope

- Add an explicit optional LoRA-FA optimizer mode for Transformers/PEFT trainers.
- Construct it with `create_lorafa_optimizer()` using the active adapter's rank,
  alpha, learning rate, and weight decay.
- Integrate scheduler construction, checkpoint/resume, distributed training, and
  Soup's optimizer reporting.
- Define and test compatibility with LoRA+, DoRA, rsLoRA, quantization,
  DeepSpeed/FSDP, and layer streaming; reject unsupported combinations clearly.
- Keep MLX out of scope unless an equivalent native implementation is designed and
  measured separately.

## Acceptance criteria

- [ ] A matrices are frozen and excluded from optimizer updates; B matrices train.
- [ ] Checkpoint/resume reproduces optimizer and scheduler state.
- [ ] A small end-to-end training smoke test produces finite loss and gradients.
- [ ] Peak memory, throughput, and quality are compared with ordinary LoRA using
      the same model, data, tokens, rank, dtype, and device.
- [ ] Documentation states the measured operating point and avoids generalizing
      the toy saved-tensor result to total VRAM.

**Comment by webdevsamran** [comment]
Hi team, I would like to work on adding PEFT LoRA-FA optimizer support.

Freezing LoRA A matrices via `create_lorafa_optimizer` significantly cuts activation memory retention by avoiding backward passes on the A projection during adapter training. I can wire this into the trainer optimizer setup with an optional configuration flag, verify parameter freezing behavior on real PEFT models, and add test coverage.

Could you please assign this to me? Thanks!

**Comment by MakazhanAlpamys** [comment]
@webdevsamran — **yours, and I have labelled it `claimed` so the label says so rather than only this comment. Apologies for the day of silence; that is on me, not a hesitation about you.**

Your plan is the right one, and the timing is now good rather than merely acceptable: **#738 merged an hour ago** (`d7aea47`), and it deliberately left the seam you need. @abdulwaarith0 asked on #724 whether to fold this in and chose to keep them separate so that `attach_loraplus_optimizer` in `utils/peft_wiring.py` would be the pattern a `attach_lorafa_optimizer` sits beside. Read that function first — it answers most of the wiring questions before you hit them.

## What it already establishes, so you do not rediscover it

- **Assign to `trainer.optimizer` after the trainer exists.** `Trainer.create_optimizer` builds one only when `self.optimizer is None`, so a post-construction assignment is respected and the scheduler is still derived from it with the configured warmup.
- **Take the optimizer class and betas/eps from the run's configured optimizer** via `Trainer.get_optimizer_cls_and_kwargs`, rather than hardcoding AdamW.
- **Check PEFT's own kwarg spelling.** LoRA+ ignores a plain `weight_decay` and honours `loraplus_weight_decay`; verify what `create_lorafa_optimizer` actually accepts rather than assuming symmetry.
- **Guard the unsupported combinations loudly.** LoRA+ with GaLore, and either on a non-LoRA run, are refused with a named message.

## Three things I will ask for at review, so you can build them in rather than bolt them on

1. **A real PEFT model and a real `transformers.Trainer` in the tests, not a mock.** A `MagicMock` auto-creates the LoRA parameter groups this path depends on, so the easy version of the test passes on broken code. This is not a style preference — it is the specific reason #738's first round was trustworthy.
2. **An assertion that A is actually frozen and B actually trains** — `requires_grad` on both, after the optimizer is built, on a real model. Your issue body already reports that behaviour from PEFT 0.20.0; the test should pin it, because that freeze is the whole memory argument.
3. **The wiring-coverage scan must stay honest.** `tests/test_issue724_loraplus_wiring_coverage.py` pins the LoRA+-wired set to exactly three wrappers and keeps an earned exemption list. If LoRA-FA is wired into the same three, add the parallel coverage for it; if it lands anywhere else, expect that scan to fail until it is updated deliberately. That friction is the design, not an obstacle.

## Two scoping questions, answered up front

- **Config surface.** A new `TrainingConfig` field is fine, but it must be **mutually exclusive with `loraplus_lr_ratio`** at parse time with a named error, not silently last-one-wins — two optimizer overrides on one run is exactly the class of silent misconfiguration this repo keeps finding.
- **Which trainers.** Wire it into the same three (`sft`, `pretrain`, `embedding`) and no further. See **#745**, filed by @abdulwaarith0 an hour ago: the other thirteen PEFT-building trainers ignore `loraplus_lr_ratio` today, and widening LoRA-FA into them in the same PR would double a gap rather than close one.

The activation-memory probe in your issue body is a good measurement to keep — if you can restate it against the real wired path rather than an isolated `X @ A @ B`, that is the number the PR should lead with.

No deadline attached. If you want to check a design decision before writing tests, ask here and I will answer the same day — which, this time, I did not.

**Comment by MakazhanAlpamys** [comment]
@webdevsamran — a status check, not a nudge to hurry.

You claimed the LoRA-FA optimizer here on 2026-09-07 and there is no branch or PR against it yet. Six days is not long for work of this size, so this is only about keeping the label honest: still planning to take it?

If it has gone cold or turned out bigger than it looked, say so and I will drop the `claimed` label with no hard feelings — that is useful information, not a failure. If you are mid-way and just quiet, say that instead and I will leave it held for as long as you need.

**Comment by webdevsamran** [comment]
Hi @MakazhanAlpamys, thanks for checking in! Yes, I am actively taking this and working on it right now.

I have your review guidance mapped out:
- Mirroring `attach_loraplus_optimizer` in `utils/peft_wiring.py` with `attach_lorafa_optimizer` using `Trainer.get_optimizer_cls_and_kwargs` and exact `create_lorafa_optimizer` kwargs
- Adding mutually exclusive validation with `loraplus_lr_ratio` in `TrainingConfig`
- Wiring into the three designated trainers (`sft`, `pretrain`, `embedding`)
- Real `PeftModel` + `Trainer` test validating that A matrices are frozen (`requires_grad=False`) while B trains
- Wiring-coverage scan in the test suite
Submitting the PR shortly today!

**Comment by webdevsamran** [comment]
﻿PR is up for review at #962!

All requested criteria from your comment are implemented and tested:
- Added `training.use_lorafa` to `TrainingConfig` with parse-time mutual exclusion against `loraplus_lr_ratio` and `use_galore`.
- Implemented `attach_lorafa_optimizer` in `utils/peft_wiring.py` beside `attach_loraplus_optimizer`, passing `weight_decay` directly to PEFT's `create_lorafa_optimizer` and preserving `betas`/`eps`.
- Wired into `sft`, `pretrain`, and `embedding`.
- Added unit tests on a real `PeftModel` and `Trainer` asserting `requires_grad=False` on `lora_A` and `requires_grad=True` on `lora_B`, verifying optimizer class, weight decay, conflict guards, and save/resume state reproduction.
- Added wiring-coverage scan in `tests/test_issue725_lorafa_wiring_coverage.py` mirroring #738 and keeping the 13 un-wired trainers exempt (#745).
- All 96 LoRA+ and LoRA-FA tests pass cleanly, and ruff is clean.

**Comment by MakazhanAlpamys** [comment]
Closed by #962 (`ebdb4a7b`) from @webdevsamran — thank you.

`training.use_lorafa: true` now wires PEFT's `LoraFAOptimizer` into the SFT-family trainers: the A matrices
stay frozen and only B is trained, so the optimizer state is a fraction of plain LoRA's.

Try:

```yaml
training:
  use_lorafa: true
  lora:
    r: 16
```

What makes it safe to enable is the set of refusals around it, each pinned by its own mutation: Spectrum,
LISA, `lora.r: 0` full fine-tuning, VeRA (refused at parse **and** at runtime) and non-AdamW optimizers are
all rejected rather than silently producing an optimizer with nothing to optimize. On `backend: mlx` the
field is declared as unread in the support registry, so `soup doctor --config` will say so instead of
implying it works.

The operating point is published as `benchmarks/gate-725-lorafa-operating-point.md` with its harness
(`benchmarks/harness/issue725_lorafa_operating_point.py`), and `docs/peft-and-efficiency.md` carries the
measured/analytic split and names the one combination that is not validated (`stream_layers`).
